### PR TITLE
Model package e2e 

### PR DESCRIPTION
**Pull request by ryansteakley**
Adds ModelPackage CRD and tests for both unversioned and versioned model packages. Also includes samples for ModelPackage and ModelPackageGroup. Updates sagemaker-controller to the latest codegen and runtime changes.

Do not review/merge until ModelPackageGroup is merged into main and tests are running

### CODE REVIEW

File: test/e2e/tests/test_model_package.py
Comment (surajkota) [inline test/e2e/tests/test_model_package.py:109]: move 109 and 113 checks to the test, we only need this in fixtures where we are creating dependent resources (e.g. xgboost_model_package_group in this test) and are not the resource under test. Usually prefer failing test rather than a fixture

same for versioned

File: test/e2e/tests/test_model_package.py
Comment (surajkota) [inline test/e2e/tests/test_model_package.py:55]: add check for arn before proceeding

**Review by surajkota — CHANGES_REQUESTED** [review]
Do we need to set a terminal condition if modelPackageStatus is `Failed` like other resources? Although, I am not sure when would it reach `Failed` and how the user is given information about the error since there is no `FailureReason` field. Can you find out about this?

Clean and nicely done PR. Few comments

File: generator.yaml
Comment (surajkota) [inline generator.yaml:530]: nit: remove `sdk_` prefix for this?

File: test/e2e/tests/test_model_package.py
Comment (ryansteakley) [reply test/e2e/tests/test_model_package.py:215]: test for tags is included in modelpackage. However it only applies to unversioned modelpackage. Tags for versioned_model_package cause error and states to include such tags to the model_package_grouop

File: pkg/resource/model_package/sdk.go
Comment (ryansteakley) [reply pkg/resource/model_package/sdk.go:942]: Update doesn't exist for unversioned Modelpackage. Not sure what logic needs to be added for versioned Model case since an ARN will always exists for them when going through sdkfind. And update call would come after the create which populates the ARN

File: pkg/resource/model_package/resource.go
Comment (ryansteakley) [reply pkg/resource/model_package/resource.go:93]: Discussed on a call with Nicholas but seems to be that ModelPackageName will be going through the describe call from sdkFind and will use arn or name as already added in hookcode. So current code should already work for this case. Am currently testing adopting resource for both cases however.

File: pkg/resource/model_package/sdk.go
Comment (ryansteakley) [reply pkg/resource/model_package/sdk.go:942]: Looks like it is able to adopt both resources. Although when deleting the versioned modelpackage through kubectl it removes it from k8s side but still exists on sagemaker.

File: test/e2e/tests/test_model_package.py
Comment (surajkota) [inline test/e2e/tests/test_model_package.py:56]: nit: change to logging.error. Because of [this configuration](https://github.com/aws-controllers-k8s/sagemaker-controller/blob/522c147abcca72d97539f7296f000105b7b1dc26/test/e2e/service_bootstrap.py#L123), debug logs are not getting printed

we need to change for others as well but lets not do in this PR

File: pkg/resource/model_package/sdk.go
Comment (surajkota) [reply pkg/resource/model_package/sdk.go:942]: ok for now to leave this as-is but please create a github issue and link this conversation, basically the else block should not be generated if ARN cannot be constructed from name or throw an an error like, missingRequiredField

File: pkg/resource/model_package/resource.go
Comment (surajkota) [reply pkg/resource/model_package/resource.go:93]: Do you mean you tried specifying ARN in `spec.aws.nameOrID` to adopt for versioned model and it worked? Well, technically it might have worked because of the nature of this API accepting ARN as well as name as input to API but since customer knows versioned model package does not have a name, they will use [spec.aws.arn](https://github.com/aws-controllers-k8s/sagemaker-controller/blob/main/config/crd/common/bases/services.k8s.aws_adoptedresources.yaml#L49) instead of [spec.aws.nameOrID](https://github.com/aws-controllers-k8s/sagemaker-controller/blob/main/config/crd/common/bases/services.k8s.aws_adoptedresources.yaml#L53) so I think we should stick to using ARN for both cases. It should work out of the box

Comment (surajkota) [reply pkg/resource/model_package/resource.go:93]: > Looks like it is able to adopt both resources. Although when deleting the versioned modelpackage through kubectl it removes it from k8s side but still exists on sagemaker.

interesting, I noticed service side check is missing in test. Added a comment. Lets fix this

**Review by surajkota — CHANGES_REQUESTED** [review]

File: test/e2e/tests/test_model_package.py
Comment (surajkota) [inline test/e2e/tests/test_model_package.py:235]: nit: logging.error

File: test/e2e/tests/test_model_package.py
Comment (surajkota) [inline test/e2e/tests/test_model_package.py:227]: the objective of assert at 233 is to check from the service side, so recommend using `list_model_packages()` and filter by PackageGroupName. This should contain only one item because of name randomisation and use the ARN from that for the assertion

**Review by surajkota — APPROVED** [review]
small changes

File: generator.yaml
Comment (surajkota) [inline generator.yaml:569]: add a comment about the reason

**Comment by surajkota** [comment]
@akartsky need another review. It's almost ready to merge

File: test/e2e/tests/test_model_package.py
Comment (akartsky) [inline test/e2e/tests/test_model_package.py:34]: move this to `test/e2e/common/config.py`

Comment (surajkota) [reply test/e2e/tests/test_model_package.py:34]: other tests only doing it if it's used by multiple tests. so this is ok

File: test/e2e/resources/xgboost_versioned_model_package.yaml
Comment (ryansteakley) [reply test/e2e/resources/xgboost_versioned_model_package.yaml:36]: versioned model package cannot have tags.

File: generator.yaml
Comment (ryansteakley) [reply generator.yaml:530]: modelpackagename can either be the arn in case of Versioned ModelPackage or the modelpackagename in unversioned modelpackage. aws-sdk-go calls the field modelpackagename  and the description says "NAME OR ARN"

File: generator.yaml
Comment (ryansteakley) [reply generator.yaml:528]: they were added in Brianna's pr previously

File: generator.yaml
Comment (ryansteakley) [reply generator.yaml:526]: they were added in Brianna's pr previously

File: templates/model_package/set_resource_name_as_arn.go.tpl
Comment (ryansteakley) [reply templates/model_package/set_resource_name_as_arn.go.tpl:8]: ModelPackageName is internal input struct name for the field that takes in either an ARN or the ModelPackageName. Naming of the field is poor since ModelPackageName is an existing field which causes confusion

**Review by akartsky — APPROVED** [review]

**Review by surajkota — APPROVED** [review]
/approve
/lgtm

**Comment by surajkota** [comment]
> https://github.com/aws-controllers-k8s/sagemaker-controller/pull/54#pullrequestreview-714193265

Missed addressing this. Not blocking this PR. Lets iterate

**Pull request merged**
